### ISSUE TITLE
Logrotation of Logfiles

### ISSUE THREAD

**Issue by PrismaComputer** [hmailserver/hmailserver#105]
It's (nearly) impossible to rotate hmailservers logfiles because of it's date stamped filenames.

Using a windows implementation of logrotate needs files to be renamed first before logrotate starts. Therefore files need to be closed e.g. achieved by stopping hmailserver or keeping the files not open.

Please add a build-in log rotation.

**Comment by OutbackMatt** [comment]
I'm not sure that I understand this request...
I use a set of VB scripts and 7Zip to compress log files each month, and the combine all of the smaller log files (eg Error logs) into a single text file for the month. Log file names can easily be renamed after the day has finished, even via script...

What is stopping you from doing something similar?

Why do you specifically want logrotate included into hMailserver. What advantage are trying to obtain

**Comment by PrismaComputer** [comment]
Yes of course, I can use a script for backing up hmailserver while a build-in backup exists, but with limitations. And yes I can use a pullmail/fetchmail script to fetch mail from pop3, while a build-in pop3c exists, but with limitations. And yes I can use a script to rotate the logs while a daily truncation is already done, but no clean up of old files. Scripting I a beautiful thing!

Good idea to use the date and zip all logs not from today. And after that you have to rotate the zip by yourself. I'm pretty sure there are many possibilities to that by a script.

But does it make sense to script basic functionality while a built-in functionality is implemented incomplete? On a *nix machine I don't expect the services to rotate their logfiles by themselves, but I expect them to write the files in an order to be able to rotate them with standard procedures.

**Comment by OutbackMatt** [comment]
But Windows doesn't have 'standard procedures' for rotating log files. You are talking about Windows making behave like *nix with a third party tool, and then getting hmailserver to work with that third party tool.

**Comment by PrismaComputer** [comment]
The BTW was a little "off topic", doesn't change the core meaning of my answer. But thank you for quoting me: Precisely because windows has no "standard procedure" to rotate a logfile it's not impertinent to expect the application to do it ;)

**Comment by visualperception** [comment]
But windows has a task scheduler for preceisely this sort of thing. All you need is some simple VBS file renaming code and then schedule it to run after midnight, say for example 01:05
It is expected that you have understanding of running a server which includes some scripting knowledge otherwise you will be constantly asking for functionality which most people running a server are capable of doing themselves. You can change it work weekly or monthly if you like. 

Set WshShell = Wscript.CreateObject("Wscript.Shell")
Set oApp = CreateObject("hMailServer.Application")
Call oApp.Authenticate("Administrator", "password")  ' ## set password here
Set oLog = oApp.Settings.Logging
Set oFSO = CreateObject("Scripting.FileSystemObject")
LogDate = DateAdd("d", -0, Now)
LogDate = Replace (LogDate, " ", "_")
LogDate = Replace (LogDate, ":", ".")
LogDate = Left(LogDate,(Len(LogDate) - 9))
OldName = oLog.CurrentEventLog
NewName = Left(oldName,(Len(OldName) - 4)) + "_" + LogDate + ".log"
If (oFSO.FileExists(OldName)) Then
 Call oApp.Stop
 oFSO.MoveFile OldName , NewName
 Call oApp.Start
End If 
Set WshShell = Nothing

**Comment by PrismaComputer** [comment]
Thanks for the answer, but I'm not sure what you want to achieve with this script. The script tries to rename a file named "hmailserver_events.log" to a file extended with date.
1. The file does not exist. Not within our setup.
2. I would be happy to have logfiles with a solid name. But the standard file names like "hmailserver_2015-09-28.log" or "ERROR_hmailserver_2015-09-28.log" are not that easy to fetch with an external rotator. Script or application anyway.
3. The script does not work with every date-time fomat, but that's not the point. If this script would be helpful, I'd be able to fix this myself.

To be more concrete, the request is: I want to configure to write logs continuously without date extension or to be able to let hmailserver rotate the logs by itself.

For your understanding: Rotation means for me zipping the old log and keep the last X zipped logs. It means not for me to have a endless list of logfiles extended with a date (what your script would achieve). This is like hmailserver works at the moment and I'd never wrote this request, if this would help.

**Comment by visualperception** [comment]
DOS/Windows copy command allows you to append one or many files together resulting in one file. These are elemenatry batch script functions which any server administrator should know how to do.
There are a myriad of possibilities for how people would want to process their log files. Your request is specific to you and you should be able to resolve it using an extremely simple DOS/Windows BAT file with a copy command.
The point I'm making is really that there is no requirement for hmailserver to do it for you but thats just my opinion.

**Comment by PrismaComputer** [comment]
It's no easy batch script to at least rotate a logfile and keep only X copys . Once rotation is done zipping by a batch is of course no problem. With your suggestions a service interruption is also needed as long hmailserver keeps the file open.

Admins used to deal with linux servers know what I'm talking about and what rotation means.

**Comment by martinknafve** [comment]
No work will be done on this in hMailServer 5.6. hMailServer 6.x will have a more flexible logging and will support for example NLog which in turn supports time-based and size-based archival.
